src/gpio.rs

```diff
@@ -413,6 +413,7 @@ macro_rules! impl_input_only {
             }
 
             pub fn into_input(self) -> Result<$pxi<Input>, EspError> {
+                esp_result!(unsafe { gpio_reset_pin($pxi::<MODE>::runtime_pin()) }, ())?;
```

Comment:
I haven't tried it, but can you use `self.reset` if you declare the signature as `pub fn into_input(mut self) -> Result<$pxi<Input>, EspError> {`?

Reply:
I believe that should work. The less invasive change would probably be to change `reset`, as we define the trait it originates from (`Pin`) and it doesn't have any real reason to be `&mut`.

Reply:
I think we can keep `Pin`'s `reset()` taking `&mut` because our `into_*` methods are taking `self`. Changing the mutability of an owned object doesn't actually change the function signature which is pretty cool! Demo of what I mean here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d69f40daa82052b61841c20fb8ce7f35

Reply:
Interesting, thanks for sharing!

I guess my argument for keeping this PR as-is would simply be that there's not really any reason for `reset()` to be `&mut` regardless, is there? It's not mutating the objects which implement `Pin` in any way, so it just seems unnecessary. Not against reverting though, if we feel that's the better option.

Reply:
It is not mutating the Rust wrapper objects, but *logically* - it does mutate the pin, in that it changes its state.
So I think it is better if it this is reflected with `&mut` in the function signature IMO.

Basically the same as all other "mutating" methods, like `set_low`, `set_high` and so on.